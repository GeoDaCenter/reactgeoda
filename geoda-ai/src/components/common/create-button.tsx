import {Button} from '@nextui-org/react';
import {ReactNode} from 'react';

export function CreateButton({
  onClick,
  isDisabled,
  children
}: {
  onClick: () => void;
  isDisabled: boolean;
  children: ReactNode;
}) {
  return (
    <Button
      radius="sm"
      color="danger"
      className="bg-rose-900"
      onClick={onClick}
      isDisabled={isDisabled}
    >
      {children}
    </Button>
  );
}
