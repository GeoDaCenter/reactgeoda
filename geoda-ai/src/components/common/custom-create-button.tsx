import {Button} from '@nextui-org/react';
import Typewriter from 'typewriter-effect';

import {GreenCheckIcon} from '../icons/green-check';
import {HeartIcon} from '../icons/heart';

export function CustomCreateButton({
  onClick,
  hide,
  label
}: {
  onClick: () => void;
  hide: boolean;
  label: string;
}) {
  return (
    <Button
      radius="full"
      className="mt-2 bg-gradient-to-tr from-pink-500 to-yellow-500 text-white shadow-none"
      onClick={onClick}
      startContent={hide ? <GreenCheckIcon /> : <HeartIcon />}
      isDisabled={hide}
    >
      <Typewriter
        options={{
          strings: label,
          autoStart: true,
          loop: false,
          delay: 10
        }}
      />
    </Button>
  );
}
