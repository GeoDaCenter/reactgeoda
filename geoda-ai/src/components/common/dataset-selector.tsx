import {datasetsSelector} from '@/store/selectors';
import {Autocomplete, AutocompleteItem} from '@nextui-org/react';
import {Key, useState} from 'react';
import {useSelector} from 'react-redux';

type DatasetSelectorProps = {
  datasetId?: string;
  setDatasetId: (datasetName: string) => void;
  label?: string;
  size?: 'sm' | 'md' | 'lg';
  optional?: boolean; // for optional vars
  isInvalid?: boolean;
};

export function DatasetSelector(props: DatasetSelectorProps) {
  const datasets = useSelector(datasetsSelector);

  const [dataId, setDataId] = useState<string>(props.datasetId || '');

  const onDatasetSelectionChange = (value: Key) => {
    const selectValue = value as string;
    props.setDatasetId(selectValue);
    // update variable in state
    setDataId(selectValue);
  };

  return (
    <Autocomplete
      label={props.label || 'Select dataset'}
      className="max-w"
      onSelectionChange={onDatasetSelectionChange}
      size={props.size || 'md'}
      selectedKey={dataId}
      isInvalid={props.isInvalid}
    >
      {datasets.map((d, i) => (
        <AutocompleteItem key={d.dataId || i} value={d.dataId}>
          {d.fileName}
        </AutocompleteItem>
      ))}
    </Autocomplete>
  );
}
