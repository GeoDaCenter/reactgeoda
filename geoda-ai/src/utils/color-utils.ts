import {COLOR_RANGES, ColorRange} from '@kepler.gl/constants';
import {HexColor} from '@kepler.gl/types';
import colorbrewer from 'colorbrewer';
import interpolate from 'color-interpolate';

const MAX_COLOR_RANGE_LENGTH = 20;

export function getDefaultColorRange(numberOfColors: number) {
  return ALL_COLOR_RANGES.find(colorRange => {
    return (
      colorRange.colors.length ===
      (numberOfColors > MAX_COLOR_RANGE_LENGTH ? MAX_COLOR_RANGE_LENGTH : numberOfColors)
    );
  });
}

export function findColorRange(newStep: number, oldColorRange?: ColorRange) {
  const newColorRange = ALL_COLOR_RANGES.find(
    colorRange =>
      colorRange.colors.length === newStep &&
      colorRange.type === oldColorRange?.type &&
      colorRange.category === oldColorRange?.category &&
      // compare the name before -[number] part
      colorRange.name?.split('-')[0] === oldColorRange?.name?.split('-')[0]
  );
  return newColorRange ?? getDefaultColorRange(newStep);
}

// @ts-ignore - colorbrewer is not typed
const colorBrewerMap: {[key: string]; string} = Object.entries(colorbrewer.schemeGroups).reduce(
  (accu, [type, palettes]: [string, any]) => ({
    ...accu,
    ...palettes.reduce(
      // @ts-ignore - colorbrewer is not typed
      (group, name) => ({
        ...group,
        [name]: type
      }),
      {}
    )
  }),
  {}
);

// convert HexColor string to rgb array
export function hexToRgb(hex: string): number[] {
  return [1, 3, 5].map(i => parseInt(hex.slice(i, i + 2), 16));
}

export function updateColorRanges(): ColorRange[] {
  // update the COLOR_RANGES array by adding missing color ranges with length of colors that are not in the array
  const colorRanges = COLOR_RANGES;
  const missingColorStartIndex = 13; // colorbrewer schemes have at most 12 colors
  for (let i = missingColorStartIndex; i <= MAX_COLOR_RANGE_LENGTH; i++) {
    for (const [keyName, colorScheme] of Object.entries(colorbrewer)) {
      if (keyName !== 'schemeGroups') {
        const originalColors = colorScheme[8]; // all colorbrewer schemes have at least 8 colors
        const interp = interpolate(originalColors);
        const colors = Array.from({length: i}, (_, j) => interp(j / (i - 1)));
        // convert colors from 'rgb(255, 255, 255)' to '#ffffff'
        const hexColors = colors.map(color => {
          const rgb = color.match(/\d+/g);
          return `#${rgb?.map(c => parseInt(c).toString(16).padStart(2, '0')).join('')}`;
        });
        colorRanges.push({
          name: `ColorBrewer ${keyName}-${i}`,
          type: colorBrewerMap[keyName],
          category: 'ColorBrewer',
          colors: hexColors as HexColor[]
        });
      }
    }
  }
  return colorRanges;
}

export const ALL_COLOR_RANGES = updateColorRanges();
