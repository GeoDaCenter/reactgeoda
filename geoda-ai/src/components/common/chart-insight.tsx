import {Button, Tooltip} from '@nextui-org/react';
import React from 'react';
import {useDispatch} from 'react-redux';
import {Icon} from '@iconify/react';

import {takeSnapshot} from '../../utils/ui-utils';
import {setDefaultPromptText, setScreenCaptured} from '../../actions';

export function ChartInsightButton({parentElementId}: {parentElementId: string}) {
  const dispatch = useDispatch();

  const getChartInsights = async () => {
    // get screenshot from the parent element
    const chartElement = document.getElementById(parentElementId);
    if (chartElement) {
      const screenshot = await takeSnapshot(chartElement);
      // dispatch the screenshot to the AI
      dispatch(setScreenCaptured(screenshot));
      // dispatch to set default prompt message in chat panel
      dispatch(setDefaultPromptText('Can you get some insights from this chart?'));
    }
  };

  return (
    <Tooltip content="Get AI Insights" size="sm">
      <Button
        className="absolute right-1 top-1 z-10"
        isIconOnly={true}
        radius="full"
        size="sm"
        color="primary"
        variant="light"
        onClick={getChartInsights}
      >
        <Icon icon="ri:gemini-fill" width={18} />
      </Button>
    </Tooltip>
  );
}
