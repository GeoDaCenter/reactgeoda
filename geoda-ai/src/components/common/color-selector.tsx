import React, {useState} from 'react';
import {Card, CardBody, Select, SelectItem} from '@nextui-org/react';
import {ColorRange} from '@kepler.gl/constants';
import {ColorPalette} from '@kepler.gl/components';
import {ALL_COLOR_RANGES, UNIQUE_COLOR_TYPES} from '@/utils/color-utils';

type ColorSelectorProps = {
  numberOfColors?: number;
  colorType?: string;
  defaultColorRange?: string;
  onSelectColorRange: (colorRange: ColorRange) => void;
};

/**
 * ColorSelector component for mapping
 * @param {number} numberOfColors - number of colors to display
 * @param {string} colorType - color type to filter
 * @param {string} defaultColorRange - default color range to select
 * @param {function} onSelectColorRange - callback function to handle color range selection
 * @returns {React.ReactNode} - ColorSelector component
 */
export function ColorSelector({
  numberOfColors = 5,
  colorType,
  defaultColorRange,
  onSelectColorRange
}: ColorSelectorProps) {
  const [selectedColorType, setSelectedColorType] = useState(colorType);

  // filter color ranges by number of colors and color type
  const colorRanges = ALL_COLOR_RANGES.filter(colorRange => {
    return (
      colorRange.colors.length === numberOfColors &&
      // filter by color type if provided
      (selectedColorType ? colorRange.type === selectedColorType : true)
    );
  });

  // find the default color range
  const selectedColorRange =
    colorRanges.find(colorRange => colorRange.name === defaultColorRange) || colorRanges[0];

  // handle color scheme selection change
  const onColorSchemeSelectionChange = (
    keys: 'all' | (Set<React.Key> & {anchorKey?: string; currentKey?: string})
  ) => {
    if (keys === 'all') return;
    const selectValue = keys.currentKey;
    const selectedColorRange = colorRanges.find(colorRange => colorRange.name === selectValue);
    if (selectedColorRange) {
      onSelectColorRange(selectedColorRange);
    }
  };

  // handle color type selection change
  const onColorTypeSelectionChange = (
    keys: 'all' | (Set<React.Key> & {anchorKey?: string; currentKey?: string})
  ) => {
    if (keys === 'all') return;
    const selectValue = keys.currentKey;
    setSelectedColorType(selectValue);
    // trigger color scheme selection change
    const selectedColorRange = colorRanges.find(colorRange => colorRange.type === selectValue);
    if (selectedColorRange) {
      onSelectColorRange(selectedColorRange);
    }
  };

  return selectedColorRange ? (
    <Card className="w-full bg-default-100">
      <CardBody>
        <span className="mb-2 text-tiny">Color Scheme</span>
        <Select
          label="Color Type"
          className="max-w"
          size="sm"
          selectedKeys={selectedColorType ? [`${selectedColorType}`] : []}
          onSelectionChange={onColorTypeSelectionChange}
          aria-label="Select color type"
        >
          {UNIQUE_COLOR_TYPES.map(_colorType => (
            <SelectItem key={`${_colorType}`} textValue={_colorType}>
              {_colorType}
            </SelectItem>
          ))}
        </Select>
        <Select
          label=""
          className="max-w"
          items={colorRanges}
          renderValue={items => {
            return items.map(item => (
              <ColorPalette
                key={item.data?.name}
                colors={item.data?.colors || []}
                isReversed={false}
                isSelected={item.data?.name === selectedColorRange.name}
              />
            ));
          }}
          selectedKeys={[`${selectedColorRange.name}`]}
          onSelectionChange={onColorSchemeSelectionChange}
          aria-label="Select color scheme"
        >
          {colorRange => (
            <SelectItem
              key={`${colorRange.name}`}
              className="gap-0 py-0"
              textValue={colorRange.name}
            >
              <ColorPalette
                colors={colorRange.colors}
                isReversed={false}
                isSelected={colorRange.name === selectedColorRange.name}
              />
            </SelectItem>
          )}
        </Select>
      </CardBody>
    </Card>
  ) : null;
}
