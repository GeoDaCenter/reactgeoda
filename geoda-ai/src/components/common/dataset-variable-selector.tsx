import {selectKeplerDataset, datasetsSelector} from '@/store/selectors';
import {getFieldNames, VARIABLE_TYPE} from '@/utils/data-utils';
import {Autocomplete, AutocompleteItem} from '@nextui-org/react';
import {Key, useMemo} from 'react';
import {useSelector} from 'react-redux';

export type onDatasetVariableSelectionChangeProps = {
  dataId?: string;
  variable?: string;
};

type DatasetVariableSelectorProps = {
  datasetId: string;
  setDatasetId: (datasetId: string) => void;
  variable: string;
  setVariable: (variable: string) => void;
  variableType?: VARIABLE_TYPE;
  onSelectionChange: ({dataId, variable}: onDatasetVariableSelectionChangeProps) => void;
  datasetLabel?: string;
  variableLabel?: string;
  size?: 'sm' | 'md' | 'lg';
};

export function DatasetVariableSelector({
  datasetId,
  setDatasetId,
  variable,
  setVariable,
  variableType = VARIABLE_TYPE.Numeric,
  onSelectionChange,
  datasetLabel,
  variableLabel,
  size
}: DatasetVariableSelectorProps) {
  const datasets = useSelector(datasetsSelector);

  const keplerDataset = useSelector(selectKeplerDataset(datasetId));

  // get numeric columns
  const columnNames = useMemo(
    () => getFieldNames(keplerDataset, variableType),
    [keplerDataset, variableType]
  );

  // handle variable change
  const onVariableSelectionChange = (value: Key | null) => {
    if (!value) return;
    const selectValue = value as string;
    onSelectionChange({variable: selectValue, dataId: keplerDataset?.id});
    setVariable(selectValue);
  };

  const onDatasetSelectionChange = (value: Key | null) => {
    if (!value) return;
    const selectValue = value as string;
    onSelectionChange({dataId: selectValue, variable});
    setDatasetId(selectValue);
  };

  return (
    <div className="flex flex-col gap-2">
      <Autocomplete
        label={datasetLabel || 'Select a dataset'}
        className="max-w"
        onSelectionChange={onDatasetSelectionChange}
        size={size || 'md'}
        selectedKey={keplerDataset.id}
        isInvalid={!keplerDataset || keplerDataset.id.length === 0}
      >
        {datasets?.map((d, i) => (
          <AutocompleteItem key={d.dataId || i} value={d.dataId}>
            {d.fileName}
          </AutocompleteItem>
        ))}
      </Autocomplete>
      <Autocomplete
        label={variableLabel || 'Select a variable'}
        className="max-w"
        onSelectionChange={onVariableSelectionChange}
        size={size || 'md'}
        selectedKey={variable}
        isInvalid={!variable || variable.length === 0}
      >
        {columnNames.map(column => (
          <AutocompleteItem key={column} value={column}>
            {column}
          </AutocompleteItem>
        ))}
      </Autocomplete>
    </div>
  );
}
