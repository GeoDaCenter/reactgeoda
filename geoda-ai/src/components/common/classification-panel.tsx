// This component will be used to display the classification panel for data or map classification

import {MappingTypes} from '@/constants';
import {useMemo, useState} from 'react';
import {ColorSelector, getDefaultColorRange} from './color-selector';
import {Select, SelectItem, Tab, Tabs} from '@nextui-org/react';
import {ColorRange} from '@kepler.gl/constants';
import {getLayer, getNumericFieldNames} from '@/utils/data-utils';
import {useSelector} from 'react-redux';
import {GeoDaState} from '@/store';
import {RateValueComponent} from './rate-component';

export const ClassificationTypes = [
  {
    label: 'Quantile',
    value: MappingTypes.QUANTILE
  },
  {
    label: 'Natural Breaks',
    value: MappingTypes.NATURAL_BREAK
  },
  {
    label: 'Equal Interval',
    value: MappingTypes.EQUAL_INTERVAL
  },
  {
    label: 'Percentile',
    value: MappingTypes.PERCENTILE
  },
  {
    label: 'Box (Hinge=1.5)',
    value: MappingTypes.BOX_MAP_15
  },
  {
    label: 'Box (Hinge=3.0)',
    value: MappingTypes.BOX_MAP_30
  },
  {
    label: 'Standard Deviation',
    value: MappingTypes.STD_MAP
  },
  {
    label: 'Unique Values',
    value: MappingTypes.UNIQUE_VALUES
  },
  {
    label: 'Co-location',
    value: MappingTypes.COLOCATION,
    disabled: true
  }
];

// create an array from 2 to 20
const DefaultNumberOfCategories = Array.from({length: 19}, (_, i) => i + 2).map(i => ({
  label: `${i}`,
  value: i
}));

export type ClassificationOnValuesChange = {
  method: string;
  variable: string;
  k: number;
  colorRange?: ColorRange;
  rateValues?: number[];
};

export type ClassificationPanelProps = {
  onValuesChange?: ({method, variable, k, colorRange}: ClassificationOnValuesChange) => void;
};

// It will be used in the Map Panel component and Bubble Chart component
export function ClassificationPanel({onValuesChange}: ClassificationPanelProps) {
  // useSelector to get layer from redux store
  const layer = useSelector((state: GeoDaState) => getLayer(state));

  // useState for number of bins
  const [k, setK] = useState(5);

  // useState for variable name
  const [variable, setVariable] = useState('');

  // useState for mapping type
  const [mappingType, setMappingType] = useState(MappingTypes.QUANTILE);

  // useState for selected color range
  const [selectedColorRange, setSelectedColorRange] = useState(getDefaultColorRange(k));

  // get numeric columns from redux store
  const numericColumns = useMemo(() => {
    const fieldNames = getNumericFieldNames(layer);
    return fieldNames;
  }, [layer]);

  // handle map type change
  const onMapTypeChange = (value: any) => {
    const selectValue = value.currentKey;
    setMappingType(selectValue);
    onValuesChange?.({method: selectValue, variable, k, colorRange: selectedColorRange});
  };

  // handle number of bins change
  const onKSelectionChange = (value: any) => {
    const kValue = Number(value.currentKey);
    setK(kValue);
    onValuesChange?.({method: mappingType, variable, k: kValue, colorRange: selectedColorRange});
  };

  // handle variable change
  const onVariableSelectionChange = (value: any) => {
    const selectValue = value.currentKey;
    setVariable(selectValue);
    onValuesChange?.({
      method: mappingType,
      variable: selectValue,
      k,
      colorRange: selectedColorRange
    });
  };

  // handle color range selection change
  const onSelectColorRange = (p: ColorRange) => {
    setSelectedColorRange(p);
    onValuesChange?.({method: mappingType, variable, k, colorRange: p});
  };

  // handle rate values change
  const onRateValuesChange = (values: unknown | unknown[], label: string) => {
    if (Array.isArray(values)) {
      onValuesChange?.({
        method: mappingType,
        variable: label,
        k,
        colorRange: selectedColorRange,
        rateValues: values
      });
    }
  };

  return (
    <div className="flex flex-col gap-2">
      <Tabs key="classification-panel-tabs" variant="underlined" aria-label="classification-tabs">
        <Tab key="basic-mapping" title="Basic Mapping">
          <Select label="Variable" className="max-w" onSelectionChange={onVariableSelectionChange}>
            {numericColumns.map((col: string) => (
              <SelectItem key={col} value={col}>
                {col}
              </SelectItem>
            ))}
          </Select>
        </Tab>
        <Tab key="rate-mapping" title="Rate Mapping">
          <RateValueComponent onValuesChange={onRateValuesChange} />
        </Tab>
      </Tabs>
      <Select
        label="Classification Method"
        className="max-w"
        onSelectionChange={onMapTypeChange}
        defaultSelectedKeys={[mappingType]}
        disabledKeys={ClassificationTypes.filter(t => t.disabled).map(t => t.value)}
      >
        {ClassificationTypes.map(mappingType => (
          <SelectItem key={mappingType.value} value={mappingType.value}>
            {mappingType.label}
          </SelectItem>
        ))}
      </Select>
      <Select
        label="Number of Categories"
        className="max-w"
        onSelectionChange={onKSelectionChange}
        defaultSelectedKeys={[`${k}`]}
      >
        {DefaultNumberOfCategories.map(bin => (
          <SelectItem key={bin.value} value={bin.value}>
            {bin.label}
          </SelectItem>
        ))}
      </Select>
      <ColorSelector
        numberOfColors={k}
        defaultColorRange={selectedColorRange?.name}
        onSelectColorRange={onSelectColorRange}
      />
    </div>
  );
}
