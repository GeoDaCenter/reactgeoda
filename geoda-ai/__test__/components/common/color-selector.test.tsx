import React from 'react';
import {render, fireEvent} from '@testing-library/react';
import {ColorSelector} from '@/components/common/color-selector';
import {ALL_COLOR_RANGES} from '@/utils/color-utils';

describe('ColorSelector', () => {
  const mockOnSelectColorRange = jest.fn();

  beforeEach(() => {
    mockOnSelectColorRange.mockClear();
  });

  test('renders with default props', () => {
    const {container} = render(<ColorSelector onSelectColorRange={mockOnSelectColorRange} />);
    expect(container).toMatchSnapshot();
  });

  test('filters color ranges based on numberOfColors prop', () => {
    const numberOfColors = 3;
    const {container} = render(
      <ColorSelector numberOfColors={numberOfColors} onSelectColorRange={mockOnSelectColorRange} />
    );

    const filteredRanges = ALL_COLOR_RANGES.filter(range => range.colors.length === numberOfColors);

    // Verify that only color ranges with specified number of colors are rendered
    filteredRanges.forEach(range => {
      expect(container.innerHTML).toContain(range.name);
    });
  });

  test('uses default color range when specified', () => {
    const defaultColorRange = ALL_COLOR_RANGES[0].name;
    const {container} = render(
      <ColorSelector
        defaultColorRange={defaultColorRange}
        onSelectColorRange={mockOnSelectColorRange}
      />
    );

    expect(container.innerHTML).toContain(defaultColorRange);
  });

  test.skip('calls onSelectColorRange when selection changes', async () => {
    const {getByRole} = render(<ColorSelector onSelectColorRange={mockOnSelectColorRange} />);

    const select = getByRole('combobox');
    fireEvent.click(select);

    // Select the first option
    const options = document.querySelectorAll('[role="option"]');
    fireEvent.click(options[1]); // Select second option

    expect(mockOnSelectColorRange).toHaveBeenCalled();
  });

  test.skip('returns null when no color ranges match the criteria', () => {
    const {container} = render(
      <ColorSelector
        numberOfColors={999} // An impossible number of colors
        onSelectColorRange={mockOnSelectColorRange}
      />
    );

    expect(container.firstChild).toBeNull();
  });
});
