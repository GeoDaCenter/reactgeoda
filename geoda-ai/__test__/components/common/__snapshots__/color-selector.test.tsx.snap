// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorSelector renders with default props 1`] = `
<div>
  <div
    class="flex flex-col relative overflow-hidden h-auto text-foreground box-border outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-medium rounded-large transition-transform-background motion-reduce:transition-none w-full bg-default-100"
    tabindex="-1"
  >
    <div
      class="relative flex w-full p-3 flex-auto flex-col place-content-inherit align-items-inherit h-auto break-words text-left overflow-y-auto subpixel-antialiased"
    >
      <span
        class="mb-2 text-tiny"
      >
        Color Scheme
      </span>
      <div
        class="group inline-flex flex-col relative w-full transition-background motion-reduce:transition-none !duration-150 max-w"
        data-filled="true"
        data-has-label="true"
        data-has-value="true"
        data-slot="base"
      >
        <div
          aria-hidden="true"
          data-a11y-ignore="aria-hidden-focus"
          data-testid="hidden-select-container"
          style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
        >
          <input
            style="font-size: 16px;"
            tabindex="-1"
            type="text"
            value="all"
          />
          <label>
            Color Type
            <select
              size="5"
              tabindex="-1"
              title=""
            >
              <option />
              <option
                value="all"
              >
                all
              </option>
              <option
                value="diverging"
              >
                diverging
              </option>
              <option
                value="sequential"
              >
                sequential
              </option>
              <option
                value="qualitative"
              >
                qualitative
              </option>
              <option
                value="singlehue"
              >
                singlehue
              </option>
            </select>
          </label>
        </div>
        <div
          class="w-full flex flex-col"
          data-slot="mainWrapper"
        >
          <button
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-label="Select color type"
            aria-labelledby="react-aria-:r7: react-aria-:r2:"
            class="relative w-full inline-flex shadow-sm tap-highlight-transparent group-data-[focus=true]:bg-default-100 rounded-small flex-col items-start justify-center gap-0 bg-default-100 data-[hover=true]:bg-default-50 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 h-12 min-h-12 py-1.5 px-3"
            data-slot="trigger"
            id="react-aria-:r2:"
            type="button"
          >
            <label
              class="block absolute z-10 subpixel-antialiased text-foreground-500 pointer-events-none cursor-pointer will-change-auto origin-top-left rtl:origin-top-right !duration-200 !ease-out transition-[transform,color,left,opacity] motion-reduce:transition-none group-data-[filled=true]:text-default-600 group-data-[filled=true]:scale-85 text-small group-data-[filled=true]:-translate-y-[calc(50%_+_theme(fontSize.tiny)/2_-_8px)] pe-2 max-w-full text-ellipsis overflow-hidden"
              data-slot="label"
              id="react-aria-:r3:"
            >
              Color Type
            </label>
            <div
              class="inline-flex h-full w-[calc(100%_-_theme(spacing.6))] min-h-4 items-center gap-1.5 box-border group-data-[has-label=true]:pt-4"
              data-slot="innerWrapper"
            >
              <span
                class="text-foreground-500 font-normal w-full text-left rtl:text-right text-small truncate group-data-[has-value=true]:text-default-foreground"
                data-slot="value"
                id="react-aria-:r7:"
              >
                all
              </span>
            </div>
            <svg
              aria-hidden="true"
              class="absolute end-3 w-4 h-4 transition-transform duration-150 ease motion-reduce:transition-none data-[open=true]:rotate-180"
              data-slot="selectorIcon"
              fill="none"
              focusable="false"
              height="1em"
              role="presentation"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              width="1em"
            >
              <path
                d="m6 9 6 6 6-6"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="group flex flex-col w-full transition-background motion-reduce:transition-none !duration-150 group relative justify-end data-[has-label=true]:mt-[calc(theme(fontSize.small)_+_10px)] max-w"
        data-slot="base"
      >
        <div
          aria-hidden="true"
          data-a11y-ignore="aria-hidden-focus"
          data-testid="hidden-select-container"
          style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
        >
          <input
            style="font-size: 16px;"
            tabindex="-1"
            type="text"
            value="undefined"
          />
          <label>
            <select
              size="45"
              tabindex="-1"
              title=""
            >
              <option />
              <option
                value="Uber Viz Diverging 1"
              >
                Uber Viz Diverging 1
              </option>
              <option
                value="Uber Viz Sequential 3"
              >
                Uber Viz Sequential 3
              </option>
              <option
                value="Uber Viz Qualitative 1"
              >
                Uber Viz Qualitative 1
              </option>
              <option
                value="UberPool 5"
              >
                UberPool 5
              </option>
              <option
                value="Ice And Fire 5"
              >
                Ice And Fire 5
              </option>
              <option
                value="Global Warming 5"
              >
                Global Warming 5
              </option>
              <option
                value="Sunrise"
              >
                Sunrise
              </option>
              <option
                value="Ocean Green 5"
              >
                Ocean Green 5
              </option>
              <option
                value="Pink Wine 5"
              >
                Pink Wine 5
              </option>
              <option
                value="Purple Blue Yellow 5"
              >
                Purple Blue Yellow 5
              </option>
              <option
                value="ColorBrewer YlGn-5"
              >
                ColorBrewer YlGn-5
              </option>
              <option
                value="ColorBrewer YlGnBu-5"
              >
                ColorBrewer YlGnBu-5
              </option>
              <option
                value="ColorBrewer GnBu-5"
              >
                ColorBrewer GnBu-5
              </option>
              <option
                value="ColorBrewer BuGn-5"
              >
                ColorBrewer BuGn-5
              </option>
              <option
                value="ColorBrewer PuBuGn-5"
              >
                ColorBrewer PuBuGn-5
              </option>
              <option
                value="ColorBrewer PuBu-5"
              >
                ColorBrewer PuBu-5
              </option>
              <option
                value="ColorBrewer BuPu-5"
              >
                ColorBrewer BuPu-5
              </option>
              <option
                value="ColorBrewer RdPu-5"
              >
                ColorBrewer RdPu-5
              </option>
              <option
                value="ColorBrewer PuRd-5"
              >
                ColorBrewer PuRd-5
              </option>
              <option
                value="ColorBrewer OrRd-5"
              >
                ColorBrewer OrRd-5
              </option>
              <option
                value="ColorBrewer YlOrRd-5"
              >
                ColorBrewer YlOrRd-5
              </option>
              <option
                value="ColorBrewer YlOrBr-5"
              >
                ColorBrewer YlOrBr-5
              </option>
              <option
                value="ColorBrewer Purples-5"
              >
                ColorBrewer Purples-5
              </option>
              <option
                value="ColorBrewer Blues-5"
              >
                ColorBrewer Blues-5
              </option>
              <option
                value="ColorBrewer Greens-5"
              >
                ColorBrewer Greens-5
              </option>
              <option
                value="ColorBrewer Oranges-5"
              >
                ColorBrewer Oranges-5
              </option>
              <option
                value="ColorBrewer Reds-5"
              >
                ColorBrewer Reds-5
              </option>
              <option
                value="ColorBrewer Greys-5"
              >
                ColorBrewer Greys-5
              </option>
              <option
                value="ColorBrewer PuOr-5"
              >
                ColorBrewer PuOr-5
              </option>
              <option
                value="ColorBrewer BrBG-5"
              >
                ColorBrewer BrBG-5
              </option>
              <option
                value="ColorBrewer PRGn-5"
              >
                ColorBrewer PRGn-5
              </option>
              <option
                value="ColorBrewer PiYG-5"
              >
                ColorBrewer PiYG-5
              </option>
              <option
                value="ColorBrewer RdBu-5"
              >
                ColorBrewer RdBu-5
              </option>
              <option
                value="ColorBrewer RdGy-5"
              >
                ColorBrewer RdGy-5
              </option>
              <option
                value="ColorBrewer RdYlBu-5"
              >
                ColorBrewer RdYlBu-5
              </option>
              <option
                value="ColorBrewer Spectral-5"
              >
                ColorBrewer Spectral-5
              </option>
              <option
                value="ColorBrewer RdYlGn-5"
              >
                ColorBrewer RdYlGn-5
              </option>
              <option
                value="ColorBrewer Accent-5"
              >
                ColorBrewer Accent-5
              </option>
              <option
                value="ColorBrewer Dark2-5"
              >
                ColorBrewer Dark2-5
              </option>
              <option
                value="ColorBrewer Paired-5"
              >
                ColorBrewer Paired-5
              </option>
              <option
                value="ColorBrewer Pastel1-5"
              >
                ColorBrewer Pastel1-5
              </option>
              <option
                value="ColorBrewer Pastel2-5"
              >
                ColorBrewer Pastel2-5
              </option>
              <option
                value="ColorBrewer Set1-5"
              >
                ColorBrewer Set1-5
              </option>
              <option
                value="ColorBrewer Set2-5"
              >
                ColorBrewer Set2-5
              </option>
              <option
                value="ColorBrewer Set3-5"
              >
                ColorBrewer Set3-5
              </option>
            </select>
          </label>
        </div>
        <div
          class="w-full flex flex-col"
          data-slot="mainWrapper"
        >
          <button
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-label="Select color scheme"
            aria-labelledby="react-aria-:rf: react-aria-:ra:"
            class="relative px-3 gap-3 w-full inline-flex flex-row items-center shadow-sm tap-highlight-transparent group-data-[focus=true]:bg-default-100 h-10 min-h-10 rounded-medium bg-default-100 data-[hover=true]:bg-default-50 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2"
            data-slot="trigger"
            id="react-aria-:ra:"
            type="button"
          >
            <div
              class="inline-flex h-full w-[calc(100%_-_theme(spacing.6))] min-h-4 items-center gap-1.5 box-border"
              data-slot="innerWrapper"
            >
              <span
                class="text-foreground-500 font-normal w-full text-left rtl:text-right text-small truncate group-data-[has-value=true]:text-default-foreground"
                data-slot="value"
                id="react-aria-:rf:"
              />
            </div>
            <svg
              aria-hidden="true"
              class="absolute end-3 w-4 h-4 transition-transform duration-150 ease motion-reduce:transition-none data-[open=true]:rotate-180"
              data-slot="selectorIcon"
              fill="none"
              focusable="false"
              height="1em"
              role="presentation"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              width="1em"
            >
              <path
                d="m6 9 6 6 6-6"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
